### ISSUE TITLE
remove pixel demultiply from agg_renderer

### ISSUE THREAD

**Issue by springmeyer** [mapnik/mapnik#2923]
Now that we track the premultiplied status on images I think we should consider:

 - [ ] stop calling `demultiply_alpha` at https://github.com/mapnik/mapnik/blob/master/src/agg/agg_renderer.cpp#L202
 - [ ] only demultiply during encoding when needed (like tiff supports premultipled pixes, but PNG does not, right?) 

The benefit of this would be:

 - If you want to take an image that goes through a renderer and then composite it to another image this would avoid needing to premultiply again.
 - Formats that support premultiplied pixels would avoid the speed cost of demultiply
 - Formats that do not support alpha would avoid the need for the speed cost (jpeg)
 - when needed the demultiplication might be able to be done during an existing loop over the pixels

/cc @artemp for comments

**Comment by flippmoke** [comment]
:+1:

**Comment by flippmoke** [comment]
I have been spending some time looking into this because frankly we have a big maze with the state of premultiplied in our code. In my opinion the following should always be followed:

* Only premultiply / demultiply right before the operations that need it. Upon exiting this algorithm the image should leave the image in the same state. 
* Changes in premultiplication / demultiplication state should be invisible and trivial for the user.

The state of premultiplication is already stored with the image so demultiply and premultiply operations already will not change the state! https://github.com/mapnik/mapnik/blob/master/src/image_util.cpp#L497

The reason for this line of thinking is that swapping state is something that is confusing and could be a waste of processing. Take for example `image_filter.hpp`:

* Different filter algorithms require different states of premultiplication
* Repeated algorithms of the same state could be called back to back. 
* An example https://github.com/mapnik/mapnik/blob/master/include/mapnik/image_filter.hpp#L395-L403 

So for example if I was to load a `jpg` render it with the image filter `blur` then with the filter `edge-detect` the state of the would be as follows:

 * Load Image  
 * Prepare to render (premultiply) 
 * image scaling (premultiplied required) and a new image (starting premultiplied is created)
 * filter blur (demultiply) 
 * after filter blur (premultiply) 
 * filter edge-detect (demultiply) 
 * after filter blur (premultiply) 
 * after rendering (demultiply). 
 * Save Image

So in order for this operations to be completed we did several extra steps of pre and multiplication which is just wasted CPU time and a reduction in algorithmic precision.

This premultiplication problem is also frustrating because one thing in particular, image_views. Image_views are intended to be read only rasters. Currently it is possible to save an image_view (and image) that is premultiplied as a png. In this case it gives an incorrect image because the colors are off (BUG!), we really need to fix this problem. I believe the best way is what I proposed above, however, it will require more thought in placement of premultiply and demultiply in the rendering process.

**Comment by flippmoke** [comment]
Side note:

> Formats that do not support alpha would avoid the speed cost of demultiply (jpeg)

JPEG requires a non-premultiplied datasource, otherwise the colors would be off.

**Comment by artemp** [comment]
@springmeyer @flippmoke - :+1: lets do it!

**Comment by lightmare** [comment]
I'm working on this currently -- demultiplying as needed when saving (except with palette, there I went lazy and just create a copy, demultiply and save that), and am wondering...

All the image saving methods go like this:
```c++
for (int y = 0; y < height; ++y) {
    auto const * row = image.get_row(y);
    for (int x = 0; x < width; ++x) {
        unsigned rgba = row[x];
        // bitwise decompose rgba and stuff it in format-specific buffer
```
This looks pretty awkward when the source and destination are both RGB(A). Wouldn't it be better if there was a `get_rgba8_row` method, returning pointer to raw data in `image_rgba8`, an iterator that would act like such pointer in other image types?

**Comment by artemp** [comment]
@lightmare - trying to understand ^ .  Are you saying `unsigned rgba = row[x];` is not generic and we'd better to provide `RGB(A)` iterator specialised to various formats ? Or something else ?

Slightly off-topic, but I think it would be great to revisit how we access pixels in various places. `boost::gil` has a good pointers regarding performance and scalability.

**Comment by lightmare** [comment]
@artemp Something like that. It looks kinda generic, in that it "works" with any `sizeof(pixels_type) <= sizeof(unsigned)`. But when converting e.g.: from `float` it doesn't make any sense; from gray8 one might expect to get rgba(g,g,g,1), not rgba(g,0,0,0) from the conversion.

I was thinking something really simple: defining a packed rgba8 struct, image_rgba8 would return reinterpret_cast pointer, other types would return an iterator which would hold this struct, fill and return when dereferenced. It's still more (over)work than this issue needs, though, so I'm going to defer that to a separate issue (3.1+), along with other image/view ideas I'm gathering...

For now I went with `static_assert<sizeof(pixel_type) == 4>` and direct byte access in `save_to_*` templates, they don't seem to be used with anything other than rgba8, anyway.

**Comment by artemp** [comment]
>For now I went with static_assert<sizeof(pixel_type) == 4> and direct byte access in save_to_* templates, they don't seem to be used with anything other than rgba8, anyway.

Sounds reasonable ^. I can't think of anything rather than `rgba8` .. maybe @flippmoke has some crazy ideas ?

**Comment by flippmoke** [comment]
@artemp @lightmare granted I haven't had much time to dive into all this but my general thought has been that we should convert the core of mapnik and change image objects such that we could use boost gil. I would love to be able to write generalized algorithms for image processing in many places and I wonder if we could adapt agg to use this?

**Comment by lightmare** [comment]
> change image objects such that we could use boost gil

I agree, despite the fact I'd quickly get lost, because I can't seem to remember anything I knew about it, and it's kinda convoluted for a new-comer.

**Comment by springmeyer** [comment]
Going to push this to v3.0.11 since I think its exciting and demands some good/non-rushed code review first.

**Comment by lightmare** [comment]
When I remove demultiply from renderer, some visual tests are failing just because the "actual" image is not demultiplied. How should visual test runner handle this?
 - force-demultiply before comparing images (will hide possible saving errors; on the other hand those should be covererd by unit test)
 - save "actual" to string, read and compare that

**Comment by springmeyer** [comment]
Demultiply before comparing seems good to me to keep tests passing without requiring fixture updates. Or premultiplying the fixtures although I would if the results would then not be identical matches?
